tests/cases/compiler/genericMappedTypeAcceptsInferredObjectType.ts

```diff
@@ -0,0 +1,34 @@
+interface Constraint {
+	foo: string
+}
+
+interface ExtendedConstraint extends Constraint {
+	bar: string
+}
+
+type GenericMap<T> = {
+	[P in keyof T]: T[P]
+}
+
+type OptionalGenericMap<T> = {
+	[P in keyof T]?: T[P]
+}
+
+const required = <T>(x: GenericMap<T>) => x
+const optional = <T>(x: OptionalGenericMap<T>) => x
+
+const withinConstraint = <T extends Constraint>(foo: T['foo']) => {
+	required<T>({ foo }) // no error as { foo: T['foo'] } <: GenericMap<T>
+	optional<T>({}) // no error as {} <: OptionalGenericMap<T>
+	optional<T>({ foo }) // no error as { foo: T['foo'] } <: OptionalGenericMap<T>
+}
+
+const withinExtendedConstraint = <T extends ExtendedConstraint>(foo: T['foo'], bar: T['bar']) => {
+	required<T>({ foo }) // error as { foo: T['foo'] } /<: GenericMap<T> because bar is missing
+	required<T>({ bar }) // error as { bar: T['bar'] } /<: GenericMap<T> because foo is missing
+	required<T>({ foo, bar }) // no error as { foo: T['foo'], bar: T['bar'] } <: GenericMap<T>
+	optional<T>({}) // no error as {} <: OptionalGenericMap<T>
+	optional<T>({ foo }) // no error as { foo: T['foo'] } <: OptionalGenericMap<T>
+	optional<T>({ bar }) // no error as { bar: T['bar'] } <: OptionalGenericMap<T>
```

Comment:
Strictly speaking, these two cases aren't in the scope of the original issue. As you're able to see from the baselines, the current behavior here is to throw errors. Intuitively, however, the behavior I described in the comments makes more sense to me.

I would be interested as to your thoughts on this matter 🙂 
Regarding an implementation, I wasn't able to figure out how to check whether a mapped type is optional (t.i. includes the `?`). Do we have a property that tracks that or should we check whether the value type includes `undefined`?

Reply:
The change to allow assignability for the optional case should be okay, but the other changes are not (see other comments).

I think the information about optional mapped types may be attached to the `SymbolFlags` for symbol of the key, but I'm not sure since I have not looked at that part of the typechecker before.

Reply:
For these two cases not to result in error, we would have to something like
```
isTypeAssignableTo(source, partial(objectTypeConstraint))
```
instead of `isTypeAssignableTo(source, objectTypeConstraint)`.

In other words, we'd have to check that each property of `source` is assignable to a corresponding property in `objectTypeConstraint`, but we do not want to check that each property in `objectTypeConstraint` also appears in `source`.

Are we already doing something similar somewhere in the type checker?